godot-core/src/engine/script_instance.rs

```diff
@@ -318,32 +259,186 @@ pub fn create_script_instance<T: ScriptInstance>(rs_instance: T) -> *mut c_void
+    ///
+    /// impl ScriptInstance for ExampleScriptInstance {
+    ///     type Base = Node;
+    ///
+    ///     fn call(
+    ///         this: SiMut<Self>,
+    ///         method: StringName,
+    ///         args: &[&Variant],
+    ///     ) -> Result<Variant, sys::GDExtensionCallErrorType>{
+    ///         let name = this.base().get_name();
+    ///         godot_print!("name is {name}");
+    ///         // However, we cannot call methods that require `&mut Base`, such as:
+    ///         // this.base().add_child(node);
+    ///         Ok(Variant::nil())
+    ///     }
+    ///     # fn class_name(&self) -> GString { todo!() }
+    ///     # fn set_property(_: SiMut<'_, Self>, _: godot::prelude::StringName, _: &godot::prelude::Variant) -> bool { todo!() }
+    ///     # fn get_property(&self, _: godot::prelude::StringName) -> Option<godot::prelude::Variant> { todo!() }
+    ///     # fn get_property_list(&self) -> Vec<PropertyInfo> { todo!() }
+    ///     # fn get_method_list(&self) -> Vec<MethodInfo> { todo!() }
+    ///     # fn is_placeholder(&self) -> bool { todo!() }
+    ///     # fn has_method(&self, _: godot::prelude::StringName) -> bool { todo!() }
+    ///     # fn get_script(&self) -> &Gd<Script> { todo!() }
+    ///     # fn get_property_type(&self, _: godot::prelude::StringName) -> VariantType { todo!() }
+    ///     # fn to_string(&self) -> GString { todo!() }
+    ///     # fn get_property_state(&self) -> Vec<(godot::prelude::StringName, godot::prelude::Variant)> { todo!() }
+    ///     # fn get_language(&self) -> Gd<ScriptLanguage> { todo!() }
+    ///     # fn on_refcount_decremented(&self) -> bool { todo!() }
+    ///     # fn on_refcount_incremented(&self) { todo!() }
+    ///     # fn property_get_fallback(&self, _: godot::prelude::StringName) -> Option<godot::prelude::Variant> { todo!() }
+    ///     # fn property_set_fallback(_: SiMut<'_, Self>, _: godot::prelude::StringName, _: &godot::prelude::Variant) -> bool { todo!() }
+    /// }
+    /// ```
+    pub fn base(&self) -> ScriptBaseRef<T> {
+        ScriptBaseRef::new(self.base_ref.to_gd(), self.mut_ref)
+    }
+
+    /// Returns a mutable reference suitable for calling engine methods on this object.
+    ///
+    /// This method will allow you to call back into the same object from Godot.
+    ///
+    /// ```no_run
+    /// # use godot::prelude::*;
+    /// # use godot::engine::{ScriptInstance, SiMut, ScriptLanguage, Script};
+    /// # use godot::builtin::meta::{MethodInfo, PropertyInfo};
+    /// # use godot::sys;
+    /// struct ExampleScriptInstance;
+    ///
+    /// impl ScriptInstance for ExampleScriptInstance {
+    ///     type Base = Object;
+    ///
+    ///     fn call(
+    ///         mut this: SiMut<Self>,
+    ///         method: StringName,
+    ///         args: &[&Variant],
+    ///     ) -> Result<Variant, sys::GDExtensionCallErrorType> {
+    ///         // Check whether method is available on this script
```

Comment:
Minor thing, I don't want to delay the review process futher, but this misses a full stop.

In the future, please apply inline code suggestions directly or make sure you copy-paste the whole thing. This cuts down on review cycles. Thanks! 🙂